File: tools/execute_command.py
import re

from tools.base import BaseTool, ToolContext

# The sandbox mounts ONLY the active project at /workspace (see docker-compose.yml),
# so the project root IS /workspace and other projects / the host are not reachable
# regardless of how the command is written. This in-process check is a courtesy on
# top of that hard boundary: it gives the model a clean, recoverable "outside the
# project" error when it tries to climb out with '..', instead of letting it wander
# into the throwaway container OS and hit a confusing failure.
CONTAINER_WORKDIR = "/workspace"

DESCRIPTION = (
    "Runs a shell command inside the sandbox container. "
    "The working directory IS the project root (/workspace); paths are relative to it. "
    "Parent-directory traversals ('..') are rejected — there is nothing above the "
    "project, and other projects are not mounted."
)

_TRAVERSAL_TOKEN = re.compile(r"(?:^|[\s'\"=:])\.\.(?:[/\\]|$|[\s'\"])")
_TRAVERSAL_IN_PATH = re.compile(r"[/\\]\.\.(?:[/\\]|$|[\s'\"])")


def _detect_escape(command: str) -> str | None:
    """Return an error message if the command tries to climb above the project root."""
    if _TRAVERSAL_TOKEN.search(command) or _TRAVERSAL_IN_PATH.search(command):
        return "contains a '..' path component that could escape the project directory"
    return None


class ExecuteCommandTool(BaseTool):
    parameters = {
        "type": "object",
        "properties": {
            "command": {"type": "string", "description": "The command to run"},
        },
        "required": ["command"],
    }

    def execute(self, ctx: ToolContext, /, **kwargs: object) -> str:
        command = kwargs.get("command")
        if not isinstance(command, str):
            return "Error: 'command' must be a string."
        reason = _detect_escape(command)
        if reason is not None:
            return (
                f"Error: command rejected — {reason}. "
                f"Commands run from the project root ({CONTAINER_WORKDIR}); "
                f"use paths relative to it."
            )
        result = ctx.docker.execute_command(command, workdir=CONTAINER_WORKDIR)
        return f"STDOUT: {result['stdout']}\nSTDERR: {result['stderr']}"

    def audit_metadata(self, ctx: ToolContext, /, **kwargs: object) -> dict[str, object]:
        return {"workdir": CONTAINER_WORKDIR, "project": ctx.project_name}

File: src/tools/execute-command.ts
// execute_command (V1/04) — plain shell (ls, cat, mv, mkdir, piping) inside the ROOT SANDBOX at
// /workspace. The real boundary is the Docker mount: only the active project is mounted at
// /workspace, so `..` / `$(...)` / symlinks land in the empty container OS, never in another
// project or on the host. The `..` guard below is a COURTESY — it hands a confused model a clean
// recoverable error instead of letting it wander — not the security control. NOT for language
// toolchains (npm/pip/pytest/cargo): the root sandbox is a slim base; those go to run_in_project.

import { WORKSPACE_PATH } from './context.js';
import type { JsonObject } from './json-object.type.js';
import type { ToolModule } from './tool-module.type.js';
import type { ToolResult } from './tool-result.type.js';
import { truncateHeadTail } from './truncate.js';

// Ported from tools/execute_command.py: an obvious `..` escape anywhere the command could form a
// path. Not a full shell parse (impossible + unnecessary — the mount is the real boundary).
const TRAVERSAL_TOKEN = /(?:^|[\s'"=:])\.\.(?:[/\\]|$|[\s'"])/;
const TRAVERSAL_IN_PATH = /[/\\]\.\.(?:[/\\]|$|[\s'"])/;

function looksLikeEscape(command: string): boolean {
  return TRAVERSAL_TOKEN.test(command) || TRAVERSAL_IN_PATH.test(command);
}

export const executeCommandTool: ToolModule = {
  name: 'execute_command',
  description:
    'Run a plain shell command (ls, cat, mv, mkdir, piping) in the active project at /workspace inside the sandbox. Paths are relative to the project root. NOT for language toolchains (tests, builds, npm/pip) — use run_in_project for those.',
  parameters: {
    type: 'object',
    properties: {
      command: { type: 'string', description: 'The shell command to run at /workspace.' },
    },
    required: ['command'],
  },

  async execute(ctx, args): Promise<ToolResult> {
    const metadata: JsonObject = { workdir: WORKSPACE_PATH, project: ctx.projectName };
    const command = args['command'];
    if (typeof command !== 'string') {
      return { content: { error: "'command' must be a string." }, exitStatus: -1, error: "'command' must be a string.", metadata };
    }
    if (looksLikeEscape(command)) {
      const error = "Command appears to escape /workspace (found '..'). Use paths relative to the project root.";
      return {
        content: { error, hint: 'Everything you need is under /workspace.' },
        exitStatus: -1,
        error,
        metadata,
      };
    }

    // Run in the root sandbox at /workspace. sandbox.exec is recoverable — a daemon/container
    // problem comes back as a non-zero ExecResult with the message in stderr, never a throw.
    const result = await ctx.sandbox.exec(command, { workdir: WORKSPACE_PATH });
    const combined = truncateHeadTail(`STDOUT: ${result.stdout}\nSTDERR: ${result.stderr}`);
    // Success returns the output string EVEN on a non-zero shell exit (the model needs the error
    // text); the audit row keeps the REAL exit code.
    // The exit code IS the result here, so it is the whole summary; the printer adds the elapsed time,
    // which for a shell command is the other half of what you wanted to know.
    return { content: combined, exitStatus: result.exitCode, metadata, display: { summary: `exit ${result.exitCode}` } };
  },
};
